Animate the build process with a progress bar

hartbit opened the pull request.

Yasumoto commented on Sources/Build/BuildDelegate.swift:
```diff
@@ -175,12 +176,28 @@ public final class BuildDelegate: BuildSystemDelegate {
     public var outputStream: ThreadSafeOutputByteStream
     public var isVerbose: Bool = false
     public var onCommmandFailure: (() -> Void)?
+    private var numberOfCommandsScanning = 0
```
Would wrapping these in a struct help logically group them together, or just add extra complexity?

Thinking something like

```
private struct CommandState {

  // Track progress of various commands based on their [state](https://github.com/apple/swift-llbuild/blob/a989feb2acfe9bcd14d19d1a225c71551e533b0c/products/llbuildSwift/BuildSystemBindings.swift#L300). 
  var scanningCount = 0
  var completedCount = 0
  var startedCount = 0
  var upToDateCount = 0

 var estimatedMaximum: Int {
         return completedCount + scanningCount - upToDateCount
     }
}
```

or something

hartbit replied:
Sounds like a good idea! Will implement it during the next iteration of this PR.

aciidb0mb3r replied:
I agree, we should wrap all this in a struct.

aciidb0mb3r commented on Sources/Build/BuildDelegate.swift:
```diff
@@ -250,6 +293,7 @@ public final class BuildDelegate: BuildSystemDelegate {
     }
 
     public func commandProcessHadOutput(_ command: SPMLLBuild.Command, process: ProcessHandle, data: [UInt8]) {
+        progressBar.complete(success: false)
```
This seems like a odd place to put it

aciidb0mb3r commented:
Yeah, I think it would be better to implement a clear so things don't look broken. It is pretty common to get warnings or errors during a build.

hartbit commented:
@aciidb0mb3r I've moved the command counting logic into a struct and implemented a progress bar clear function to make it clean with warnings and errors

Yasumoto commented:
Awesome 👏

hartbit commented:
A ninja-style animation was implemented following discussion in the forums https://forums.swift.org/t/swiftpm-build-animation/18108/16. @aciidb0mb3r It's now ready for review!

hartbit commented on Sources/Utility/ProgressAnimation.swift:
```diff
@@ -0,0 +1,290 @@
+import Basic
+
+/// A protocol to operate on terminal based progress animations.
+public protocol ProgressAnimationProtocol {
```
How would like me to split the work? Keep this PR using the old lit-based progress animation and in a separate PR implement the new Ninja-style animation?

hartbit commented:
@aciidb0mb3r This PR has now been rebased and is cleaner. Ready for review!

aciidb0mb3r commented:
Looks great!

aciidb0mb3r commented:
@swift-ci smoke test
